lib/plug_basic_auth.ex

```diff
@@ -38,38 +42,50 @@ defmodule PlugBasicAuth do
                            halt:            1]
 
   def init(opts) do
-    username = Keyword.fetch!(opts, :username)
-    password = Keyword.fetch!(opts, :password)
-    username <> ":" <> password
+    case Keyword.fetch(opts, :validation) do
+      {:ok, validation} ->
+        validation
+      :error ->
+        raise_undefined_validation_error(opts)
+    end
+  end
+
+  def call(conn, validation) do
+    {user, pwd} = get_credentials(conn)
+    auth_response = validation.(user, pwd)
+    respond(conn, auth_response)
   end
 
-  def call(conn, server_creds) do
+  defp get_credentials(conn) do
     conn
     |> get_auth_header
     |> parse_auth
-    |> check_creds(server_creds)
   end
 
   defp get_auth_header(conn) do
-    auth = get_req_header(conn, "authorization")
-    {conn, auth}
+    get_req_header(conn, "authorization")
   end
 
-  defp parse_auth({conn, ["Basic " <> encoded_creds | _]}) do
+  defp parse_auth(["Basic " <> encoded_creds | _]) do
     {:ok, decoded_creds} = Base.decode64(encoded_creds)
-    {conn, decoded_creds}
+    [user, pwd] = String.split(decoded_creds, ":", parts: 2)
+    {user, pwd}
   end
-  defp parse_auth({conn, _}), do: {conn, nil}
+  defp parse_auth(_), do: {nil, nil}
 
-  defp check_creds({conn, decoded_creds}, server_creds) when decoded_creds == server_creds do
-    conn
-  end
-  defp check_creds({conn, _}, _), do: respond_with_login(conn)
-
-  defp respond_with_login(conn) do
+  defp respond(conn, :authorized), do: conn
+  defp respond(conn, :unauthorized) do
     conn
     |> put_resp_header("www-authenticate", "Basic realm=\"Private Area\"")
     |> send_resp(401, "")
     |> halt
   end
+
+  defp raise_undefined_validation_error(opts) do
+    if Keyword.has_key?(opts, :username) do
+      raise ":username and :password options are deprecated and replaced by the :validation option."
```

Comment:
This seems more like a removal than a deprecation because we aren't supporting `:username` and `:password` at all anymore. Should we update this message to say something like `:username and :password options are no longer supported. Use the :validation option with a function that takes two arguments`?

Or we could consider doing an actual deprecation and still supporting `:username` and `:password` in this release and print a message at compile time letting them know to switch to `:validation`?